cpp/src/arrow/util/uri.cc

```diff
@@ -208,6 +208,12 @@ std::string Uri::path() const {
   return std::move(ss).str();
 }
 
+std::string Uri::extension() const { 
+  std::string filename = impl_->path_segments_.back().to_string();
+  size_t extension_index = filename.find(".");
```

Comment:
Should we use this instead? Though, I suppose that requires ARROW_FILESYSTEM=ON: https://github.com/apache/arrow/blob/4b74c72d1f97f35746a6aee8515639b37579f332/cpp/src/arrow/filesystem/path_util.h#L41-L43

At the very least, use `find_last_of`? https://github.com/apache/arrow/blob/4b74c72d1f97f35746a6aee8515639b37579f332/cpp/src/arrow/filesystem/path_util.cc#L82-L87

Reply:
This entire block of code is converting a Substrait LocalFiles into a FileSystemDataset so I think its safe to rely on `ARROW_FILESYSTEM` methods.

Reply:
This code is in `util/uri.cc` though, so it'll get built whether filesystems are enabled or not.